### PR TITLE
KOGITO-5642: Activate sonar check

### PR DESCRIPTION
**Pull request by jomarko**
Ensemble together with:
- https://github.com/kiegroup/kogito-pipelines/pull/253

### CODE REVIEW

**Comment by jomarko** [comment]
This PR should be ready for a review. 
- The changes were tried with a [custom job](https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/custom/job/jomarko/job/kogito-editors-java-main-pr-sonar-test/). Please notice it was configured to be triggered just for PRs with `enhancement` label. It will be triggered for all PRs after merge. The job was paired with sonarcloud can be seen [here](https://sonarcloud.io/dashboard?id=org.kie.kogito%3Akogito-editors-java&pullRequest=94)
- `.ci/jenkins/Jenkinsfile` is a definition of actual maven commands, that start the build once with coverage measurement and secondly with sonar checks
- `.ci.jenkins/dsl` folder is a convention for kiegroup, how we define pipelines for repositories. Please notice `jobs.groovy` is much more simple/shorter than `jobs.groovy` in other kiegroup repositories, e.g. kogito-runtimes. It is because for kogito-editors-java for now we want to generate just a 'pullrequest' job, i.e. job started when PR is opened.

Once we merge this ensemble, I will create jiras to fix the most severe reports from sonar.

File: pom.xml [hunk 1]
```diff
@@ -1365,6 +1365,32 @@
...
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
+
+    <profile>
+      <id>sonarcloud-analysis</id>
+      <properties>
+        <skipITs>true</skipITs>
+        <sonar.skip>true</sonar.skip>
```
Comment (jstastny-cz) [inline pom.xml:1383]: Wouldn't it make sense to have the sonar.skip property in general config (project > properties), not in a profile, and just override value true/false in particular modules as needed?

Comment (jomarko) [reply pom.xml]: Thank you for pointing to this, actually I will remove `sonarcloud-analysis` profile config completely, for now.

I know we use it for the rest of kiegroup organization, however I was not able to get it working for kogito-editors-java, I would like to proceed with command generated by soanrcloud:
```
org.sonarsource.scanner.maven:sonar-maven-plugin:sonar
```

File: pom.xml [hunk 2]
```diff
@@ -1365,6 +1365,32 @@
               <skip>${code.coverage.disabled}</skip>
             </configuration>
           </plugin>
+          <plugin>
+            <groupId>org.codehaus.mojo</groupId>
+            <artifactId>gwt-maven-plugin</artifactId>
+            <configuration>
+              <skip>true</skip>
```
[the remark this answers is not in the document]
Comment (jomarko) [reply pom.xml]: I used property of gwt plugin itself, as it is gwt prefixed, I believe the conflict with custom properties should be minimal and from my point of view we do not need to introduce new property.

**Comment by jomarko** [comment]
I created an epic with follow-up tasks for my-self. 
- https://issues.redhat.com/browse/KOGITO-5846

File: .github/workflows/sonar_analysis_kiegroup.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+name: "SonarCloud Analysis: kogito-editors-java"
```
Comment (radtriste) [inline .github/workflows/sonar_analysis_kiegroup.yml:1]: why do we have both GHAction and Jenkins for Sonar ?
I think we should focus on one or ?

File: .ci/jenkins/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            }
+            steps {
+                script {
+                    if (isSonarCloudEnabled()) {
+                        getMavenCommand('kogito-editors-java')
+                                .withOptions(['-e', '-nsu'])
+                                .withProperty('sonar.projectKey', 'org.kie.kogito:kogito-editors-java')
+                                .withProperty('sonar.organization', 'kiegroup')
+                                .withProperty('sonar.host.url', 'https://sonarcloud.io')
+                                .withProperty('sonar.login', env['SONARCLOUD_TOKEN'])
```
[the remark this answers is not in the document]
Comment (jomarko) [reply .ci/jenkins/Jenkinsfile:65]: yes, I agree with your point, the current solution is different from the rest of kiegroup repositories, where `-Psonarcloud-analysis` is used. However I was not able to get it working in this phase so decided to create this task https://issues.redhat.com/browse/KOGITO-5849

File: .ci/jenkins/dsl/jobs.groovy [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+Map getMultijobPRConfig() {
+    return [
+            parallel: true,
+            jobs : [
+                    [
+                            id: 'Kogito Tooling Editors Java',
+                            primary: true,
+                            repository: 'kogito-editors-java'
```
[the remark this answers is not in the document]
Comment (jomarko) [reply .ci/jenkins/dsl/jobs.groovy:29]: do you mean the snippet bellow?
```
repository: 'kogito-editors-java'
```

Is it safe to remove?

File: .github/workflows/sonar_analysis_kiegroup.yml (hunk 3 above)
Comment (radtriste) [reply .github/workflows/sonar_analysis_kiegroup.yml:1]: yes it should as it is the default repository used (from getJobParams)

Comment (jomarko) [reply .github/workflows/sonar_analysis_kiegroup.yml:1]: I followed approach from the optaplanner, where
- [sonarcloud.yml](https://github.com/kiegroup/optaplanner/blob/main/.github/workflows/sonarcloud.yml) defines job to check each push to main branch by sonar
- [Jenkinsfile](https://github.com/kiegroup/optaplanner/blob/main/.ci/jenkins/Jenkinsfile) defines job to check each opened PR agains main branch by sonar

Should I use just Jenkinsfiles like we do for kogito-runtimes?

File: .ci/jenkins/Jenkinsfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (radtriste) [reply .ci/jenkins/Jenkinsfile:65]: oki

File: .github/workflows/sonar_analysis_kiegroup.yml (hunk 3 above)
Comment (radtriste) [reply .github/workflows/sonar_analysis_kiegroup.yml:1]: oh my bad, I misread. so yeah it is ok :)

File: .ci/jenkins/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+import org.kie.jenkins.MavenCommand
+
+changeAuthor = env.ghprbAuthorRepoGitUrl ? util.getGroup(env.ghprbAuthorRepoGitUrl) : (env.ghprbPullAuthorLogin ?: CHANGE_AUTHOR)
+changeBranch = env.ghprbSourceBranch ?: CHANGE_BRANCH
+changeTarget = env.ghprbTargetBranch ?: CHANGE_TARGET
+
+pipeline {
+    agent {
+        label 'kie-rhel7 && kie-mem16g'
```
Comment (radtriste) [inline .ci/jenkins/Jenkinsfile:11]: do we need 16G for kie-editors-java ?

File: .ci/jenkins/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        maven 'kie-maven-3.6.3'
+        jdk 'kie-jdk11'
+    }
+    options {
+        timestamps()
+        timeout(time: getTimeoutValue(), unit: 'MINUTES')
+    }
+    environment {
+        SONARCLOUD_TOKEN = credentials('SONARCLOUD_TOKEN')
+        MAVEN_OPTS = '-Xms1024m -Xmx4g'
```
Comment (radtriste) [inline .ci/jenkins/Jenkinsfile:23]: Is that useful here ?

**Review by radtriste — APPROVED** [review]
job generation seems good apart from minor things but non blocking

**Comment by radtriste** [comment]
looks like SonarCloud is working well :D

File: .ci/jenkins/Jenkinsfile (hunk 6 above)
Comment (jomarko) [reply .ci/jenkins/Jenkinsfile]: I thought yes, however lets try 8G

https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/custom/job/jomarko/job/kogito-editors-java-main-pr-sonar-test/13/

File: .ci/jenkins/Jenkinsfile (hunk 7 above)
Comment (jomarko) [reply .ci/jenkins/Jenkinsfile]: I checked logs of the job, I didn't found it there, so removed, thank you

**Review by radtriste — APPROVED** [review]
Good for the Jenkins/DSL part

**Comment by jomarko** [comment]
@jstastny-cz @karreiro @romartin @tiagobento hello, this is just a friendly reminder. Please review when you have a moment. Or if you do not have capacity for review please let us know so we can proceed with merge.

The changes brings "new" automation to run sonar checks. "Existing" PR checks and code is not affected.

**Review by karreiro — COMMENTED** [review]
Thanks for this PR, @jomarko. I don't have a strong opinion about the usage of Sonar checks into this repository. So, I'd like to listen @romartin and @tiagobento opinion as well.

From the the code perspective, I believe we may add copyright headers into Jenkins and Groovy files (`Jenkinsfile`, `Jenkinsfile.seed`, and `jobs.groovy`). Wdyt?

Thanks again for this PR.

**Comment by jomarko** [comment]
@karreiro  thank you for sharing feedback.

_From the the code perspective, I believe we may add copyright headers into Jenkins and Groovy files (Jenkinsfile, Jenkinsfile.seed, and jobs.groovy). Wdyt?_ Actually I do not know, we do not have licences in `kogito-runtimes`:
- https://github.com/kiegroup/kogito-runtimes/tree/main/.ci/jenkins

So I propose to discuss in team generally (will send mail to bsig) and possibly fix it via separate jira for all repos.

**Review by jstastny-cz — APPROVED** [review]
LGTM, just a minor comment.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-runtime/pom.xml [hunk 8]
```diff
@@ -821,6 +821,13 @@
         </plugins>
       </build>
     </profile>
+    <profile>
+      <id>run-code-coverage</id>
+      <properties>
+        <skipITs>true</skipITs>
+        <code.coverage.disabled>true</code.coverage.disabled>
```
[the remark this answers is not in the document]
Comment (jomarko) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-runtime/pom.xml:828]: @jstastny-cz I provide some explanation in next commit, please have a look. I also checked the given module and what are current sonar findings there. See attached picture. It is a minimum comparing main dmn module, I adopted your proposal and skip the sonar now.

Screenshot from 2021-09-10 14-59-25

[the remark this answers is not in the document]
Comment (jomarko) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-runtime/pom.xml:834]: @jstastny-cz this commit https://github.com/kiegroup/kogito-editors-java/pull/94/commits/593f6baef05f910cddf05fd7bd97bf580a4a74e6, please have a look on the wording

[the remark this answers is not in the document]
Comment (jstastny-cz) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-runtime/pom.xml:834]: Looks good, thanks.

**Comment by tiagobento** [comment]
@jomarko I won't have time to review this PR. Please proceed without my review.

**Review by romartin — APPROVED** [review]

**Pull request merged**